### PR TITLE
Python pickle protocol support

### PR DESCRIPTION
**Pull request by wavefront-mike**
Customers that are using carbon-relay for replicating data between Graphite clusters have the option to send data using the Pickle protocol format (instead of the plaintext new line format).
This pull request adds support for that format.  The carbon-relay streams the data in a format that looks like:
 [Length of pickled data to follow in a 4 byte unsigned int][pickled data of the given length]<repeat ...>
See https://www.hostedgraphite.com/docs/languageguide/lg_python_pickle.html for more information.

With this pull request, customers can now use this new port in the proxy and split their data stream to go to both the Wavefront proxy and to the other Graphite cluster.  This is done by adding the Wavefront proxy (and port 5878) to the DESTINATION line in their configuration file.

I added an additional test case using test/runner.py to test a pickled stream retrieved from an existing server.

### CODE REVIEW

File: java-lib/src/main/java/com/wavefront/common/MetricMangler.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    // Store as-is; going to loop through chars anyway
+    delimiters = strDelimiters;
+  }
+
+  /**
+   * Simple struct to return both the source and the updated metric
+   * from {@link #getSourceNameFromMetric(String)}
+   */
+  public static class MetricComponents {
+    public String source;
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricMangler.java]: @Nullable

File: java-lib/src/main/java/com/wavefront/common/MetricMangler.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    public String metric;
+  }
+
+  /**
+   * Extracts the source from the metric name and returns the new metric name
+   * and the source name.
+   * @param metric the metric name
+   * @return the updated metric name and the extracted source
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricMangler.java]: missing throws IAE

File: java-lib/src/main/java/com/wavefront/common/MetricMangler.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+          buf.append('.');
+        }
+        buf.append(segment);
+      }
+    }
+    rtn.metric = buf.toString();
+
+    // Loop over host components in configured order, and replace all delimiters with dots
+    if (hostIndices != null && !hostIndices.isEmpty()) {
+      buf = new StringBuffer();
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricMangler.java]: StringBuilder (this does not need to be thread-safe)

File: java-lib/src/main/java/com/wavefront/common/MetricMangler.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    final String[] segments = metric.split("\\.");
+    final MetricComponents rtn = new MetricComponents();
+
+    // Is the metric name long enough?
+    if (segments.length < maxField) {
+      throw new IllegalArgumentException("Metric data provided was incompatible with format.");
+    }
+
+    // Assemble the newly shorn metric name, in original order
+    StringBuffer buf = new StringBuffer();
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricMangler.java]: StringBuilder

File: java-lib/src/main/java/com/wavefront/common/MetricMangler.java [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+  // Lookup set for which indices are hostname related
+  private Set<Integer> hostIndexSet = new HashSet<Integer>();
+
+  // Characters which should be interpreted as dots
+  private final String delimiters;
+
+  // Fields to remove
+  private Set<Integer> removeIndexSet = new HashSet<Integer>();
+
+  public MetricMangler(String strFields, String strDelimiters, String strFieldsToRemove) {
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricMangler.java]: Our style typically do not duplicate type into the variable name

File: java-lib/src/main/java/com/wavefront/common/MetricMangler.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+          if (iField <= 0) {
+            throw new IllegalArgumentException("Can't define a field to remove of index 0 or less; indices must be 1-based");
+          }
+          removeIndexSet.add(iField - 1);
+        }
+      }
+    }
+
+    // Store as-is; going to loop through chars anyway
+    delimiters = strDelimiters;
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricMangler.java]: this.delimiters = delimiters (see above).

File: java-lib/src/main/java/com/wavefront/common/MetricMangler.java (hunk 5 above)
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricMangler.java]: java doc to explain the fields?

Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricMangler.java]: @Nullable for fields, fieldsToRemove

Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricMangler.java]: also delimiters

File: java-lib/src/main/java/com/wavefront/common/MetricMangler.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+  // Fields to extract and assemble, in order, as the host name
+  private final List<Integer> hostIndices = new ArrayList<Integer>();
+  private int maxField = 0;
+
+  // Lookup set for which indices are hostname related
+  private Set<Integer> hostIndexSet = new HashSet<Integer>();
+
+  // Characters which should be interpreted as dots
+  private final String delimiters;
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricMangler.java]: @Nullable

File: java-lib/src/main/java/com/wavefront/common/MetricMangler.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+ *
+ */
+public class MetricMangler {
+
+  // Fields to extract and assemble, in order, as the host name
+  private final List<Integer> hostIndices = new ArrayList<Integer>();
+  private int maxField = 0;
+
+  // Lookup set for which indices are hostname related
+  private Set<Integer> hostIndexSet = new HashSet<Integer>();
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricMangler.java]: final

File: java-lib/src/main/java/com/wavefront/common/MetricMangler.java [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+  // Characters which should be interpreted as dots
+  private final String delimiters;
+
+  // Fields to remove
+  private Set<Integer> removeIndexSet = new HashSet<Integer>();
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricMangler.java]: final

File: java-lib/src/main/java/com/wavefront/common/MetricWhiteBlackList.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+import javax.annotation.Nullable;
+import java.util.regex.Pattern;
+import com.yammer.metrics.Metrics;
+import com.yammer.metrics.core.Counter;
+import com.yammer.metrics.core.MetricName;
+import org.apache.commons.lang.StringUtils;
+
+public class MetricWhiteBlackList
+{
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricWhiteBlackList.java]: line before

File: java-lib/src/main/java/com/wavefront/common/MetricWhiteBlackList.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+    this.pointLineWhiteList = StringUtils.isBlank(pointLineWhiteListRegex) ?
+        null : Pattern.compile(pointLineWhiteListRegex);
+
+    this.pointLineBlackList = StringUtils.isBlank(pointLineBlackListRegex) ?
+        null : Pattern.compile(pointLineBlackListRegex);
+
+    this.regexRejects = Metrics.newCounter(
+      new MetricName("validationRegex." + rejectCounterName, "", "points-rejected"));
+
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricWhiteBlackList.java]: blank line

File: java-lib/src/main/java/com/wavefront/common/MetricWhiteBlackList.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+  private final Pattern pointLineWhiteList;
+  @Nullable
+  private final Pattern pointLineBlackList;
+
+  private final Counter regexRejects;
+
+  public MetricWhiteBlackList(@Nullable final String pointLineWhiteListRegex,
+                              @Nullable final String pointLineBlackListRegex,
+                              final String rejectCounterName) {
+
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricWhiteBlackList.java]: blank line

File: java-lib/src/main/java/com/wavefront/common/MetricWhiteBlackList.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+      new MetricName("validationRegex." + rejectCounterName, "", "points-rejected"));
+
+  }
+
+  public boolean passes(String pointLine) {
+    if ((pointLineWhiteList != null &&
+         !pointLineWhiteList.matcher(pointLine).matches()) ||
+        (pointLineBlackList != null &&
+         pointLineBlackList.matcher(pointLine).matches()))
+    {
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricWhiteBlackList.java]: check style guide (ask any dev for our style guide to configure intellij or your ide)

File: java-lib/src/main/java/com/wavefront/ingester/PickleProtocolDecoder.java [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+/**
+ * Pickle protocol format decoder.
+ * https://docs.python.org/2/library/pickle.html
+ */
+public class PickleProtocolDecoder implements Decoder<byte[]> {
+
+  protected static final Logger logger = Logger.getLogger("agent");
+
+  private final String hostName;
+  private List<String> customSourceTags;
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/ingester/PickleProtocolDecoder.java]: final also 29

File: java-lib/src/main/java/com/wavefront/ingester/PickleProtocolDecoder.java [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+  public void decodeReportPoints(byte[] msg, List<ReportPoint> out, String customerId) {
+    InputStream is = new ByteArrayInputStream(msg);
+    Unpickler unpickler = new Unpickler();
+    Object dataRaw = null;
+    try {
+      dataRaw = unpickler.load(is);
+      if (!(dataRaw instanceof List)) {
+        throw new IllegalArgumentException("unable to unpickle data");
+      }
+
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/ingester/PickleProtocolDecoder.java]: blank line

File: java-lib/src/main/java/com/wavefront/ingester/PickleProtocolDecoder.java [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+      if (!(dataRaw instanceof List)) {
+        throw new IllegalArgumentException("unable to unpickle data");
+      }
+      
+    } catch (final java.io.IOException e) {
+      throw new IllegalArgumentException("unable to unpickle data", e);
+    }
+
+    // [(path, (timestamp, value)), ...]
+    List<Object[]> data = (List<Object[]>)dataRaw;
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/ingester/PickleProtocolDecoder.java]: spacing is off

File: java-lib/src/main/java/com/wavefront/ingester/PickleProtocolDecoder.java [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+    for (Object[] o : data) {
+      MetricMangler.MetricComponents components =
+        this.metricMangler.extractComponents(o[0].toString());
+
+      ReportPoint point = new ReportPoint();
+      Object[] details = (Object[])o[1];
+      if (details == null || details.length != 2)
+      {
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/ingester/PickleProtocolDecoder.java]: style

File: proxy/src/main/java/com/wavefront/agent/ChannelByteArrayHandler.java [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+                                 final long millisecondsPerBatch,
+                                 final int blockedPointsPerBatch,
+                                 @Nullable final String pointLineWhiteListRegex,
+                                 @Nullable final String pointLineBlackListRegex) {
+    this.decoder = decoder;
+    this.pointHandler = new PointHandler(agentAPI, daemonId, port, logLevel, validationLevel, millisecondsPerBatch, blockedPointsPerBatch);
+
+    this.prefix = prefix;
+    this.whiteBlackList = new MetricWhiteBlackList(pointLineWhiteListRegex,
+                                                   pointLineBlackListRegex,
```
Comment (panghy) [inline proxy/src/main/java/com/wavefront/agent/ChannelByteArrayHandler.java]: style

File: proxy/src/main/java/com/wavefront/agent/PushAgent.java [hunk 19]
```diff
@@ -65,6 +71,14 @@ protected void startListeners() {
         }
       }
     }
+    if (picklePorts != null) {
+        for (String strPort : picklePorts.split(",")) {
```
Comment (panghy) [inline proxy/src/main/java/com/wavefront/agent/PushAgent.java]: style indent

**Comment by panghy** [comment]
This PR is mostly good, just fix up a couple of stuff (if you're not using IntelliJ, please do... :) ) and we should be good to go

File: proxy/src/main/java/com/wavefront/agent/PushAgent.java [hunk 20]
```diff
@@ -148,11 +148,13 @@ protected void startPickleListener(String strPort, GraphiteFormatter formatter)
...
     // to the decoder.
     class FrameDecoderFactoryImpl implements StreamIngester.FrameDecoderFactory {
-        @Override
-        public ChannelInboundHandler getDecoder() {
-            return new LengthFieldBasedFrameDecoder(ByteOrder.BIG_ENDIAN, 1000000, 0, 4, 0, 4, false);
-        }
+      @Override
+      public ChannelInboundHandler getDecoder() {
+        return new LengthFieldBasedFrameDecoder(ByteOrder.BIG_ENDIAN, 1000000,
+                                                0, 4, 0, 4, false);
```
Comment (panghy) [inline proxy/src/main/java/com/wavefront/agent/PushAgent.java]: this should not align (method invocation params should not align across lines)

**Comment by panghy** [comment]
We are not exactly using google style so ideally i'd still use intellij which understands the coding style xml that we have in the repo but you're very close. I wouldn't be surprised however, if running reformat in intellij still changes the style (the reason why we want to have a single coding style so that each change doesn't introduce style "bounces")

**Comment by wavefront-mike** [comment]
Is this ready to merge?  Any other changes I should make?

**Comment by panghy** [comment]
Please do a merge and I'll take one final look. Was trying to get other things in first given the need to get a release out last week.

File: java-lib/src/main/java/com/wavefront/common/MetricMangler.java [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+ *   - source name:
+ *     - extracted from one or more components of the metric name
+ *       (where each component is separated by a '.')
+ *     - allow characters to be optionally replaced in the components extracted
+ *       as source name with '.'
+ *   - metric name:
+ *     - remove components (in addition to the source name)
+ *       (this allows things like 'hosts.sjc1234.cpu.loadavg.1m' to get
+ *        changed to cpu.loadavg.1m after extracting sjc1234)
+ *
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricMangler.java]: Add author

File: java-lib/src/main/java/com/wavefront/common/MetricMangler.java [hunk 22]
```diff
@@ -0,0 +1,155 @@
...
+   * @param removeFields comma separated field index(es) (1-based) of fields to
+   *                     remove from the metric name
+   * @throws IllegalArgumentException when one of the field index is <= 0
+   */
+  public MetricMangler(@Nullable String sourceFields,
+                       @Nullable String delimiters,
+                       @Nullable String removeFields) {
+    if (sourceFields != null) {
+      // Store ordered field indices and lookup set
+      for (String field : sourceFields.split(",")) {
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricMangler.java]: Use Splitter.on(",").omitEmptyStrings().trimResults()

File: java-lib/src/main/java/com/wavefront/common/MetricMangler.java [hunk 23]
```diff
@@ -0,0 +1,155 @@
...
+          hostIndexSet.add(fieldIndex - 1);
+          if (fieldIndex > maxField) {
+            maxField = fieldIndex;
+          }
+        }
+      }
+    }
+
+    if (removeFields != null) {
+      for (String field : removeFields.split(",")) {
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricMangler.java]: ditto

File: java-lib/src/main/java/com/wavefront/common/MetricMangler.java [hunk 24]
```diff
@@ -0,0 +1,155 @@
...
+   * @throws IllegalArgumentException when the number of segments (split on '.')
+   *         is less than the maximum source component index
+   */
+  public MetricComponents extractComponents(final String metric) {
+    final String[] segments = metric.split("\\.");
+    final MetricComponents rtn = new MetricComponents();
+
+    // Is the metric name long enough?
+    if (segments.length < maxField) {
+      throw new IllegalArgumentException("Metric data provided was incompatible with format.");
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricMangler.java]: include input metric in exception message

File: java-lib/src/main/java/com/wavefront/common/MetricMangler.java [hunk 25]
```diff
@@ -0,0 +1,155 @@
...
+        buf.append(segmentChars);
+      }
+      rtn.source = buf.toString();
+    } else {
+      rtn.source = null;
+    }
+
+    return rtn;
+  }
+
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricMangler.java]: blank lines

File: java-lib/src/main/java/com/wavefront/common/MetricWhiteBlackList.java [hunk 26]
```diff
@@ -0,0 +1,60 @@
...
+   * Counter for number of rejected metrics.
+   */
+  private final Counter regexRejects;
+
+  /**
+   * Constructor.
+   * @param pointLineWhiteListRegex the white list regular expression.
+   * @param pointLineBlackListRegex the black list regular expression
+   * @param rejectCounterName the name for the counter (validationRegex.{rejectCounterName})
+   */
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricWhiteBlackList.java]: author

File: java-lib/src/main/java/com/wavefront/common/MetricWhiteBlackList.java [hunk 27]
```diff
@@ -0,0 +1,60 @@
...
+  public MetricWhiteBlackList(@Nullable final String pointLineWhiteListRegex,
+                              @Nullable final String pointLineBlackListRegex,
+                              final String rejectCounterName) {
+    this.pointLineWhiteList = StringUtils.isBlank(pointLineWhiteListRegex)
+      ? null : Pattern.compile(pointLineWhiteListRegex);
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricWhiteBlackList.java]: nitpick: the ternary operator is usually on the previous line

File: java-lib/src/main/java/com/wavefront/common/MetricWhiteBlackList.java [hunk 28]
```diff
@@ -0,0 +1,60 @@
...
+
+    this.pointLineBlackList = StringUtils.isBlank(pointLineBlackListRegex)
+      ? null : Pattern.compile(pointLineBlackListRegex);
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricWhiteBlackList.java]: ditto

File: java-lib/src/main/java/com/wavefront/common/MetricWhiteBlackList.java [hunk 29]
```diff
@@ -0,0 +1,60 @@
...
+
+    this.regexRejects = Metrics.newCounter(
+      new MetricName("validationRegex." + rejectCounterName, "", "points-rejected"));
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricWhiteBlackList.java]: Use TaggedMetricName

File: java-lib/src/main/java/com/wavefront/common/MetricWhiteBlackList.java [hunk 30]
```diff
@@ -0,0 +1,60 @@
...
+
+  /**
+   * Check to see if the given point line or metric passes the white and black
+   * list.
+   * @param pointLine the line to check
+   * @return true if the line passes checks; false o/w
+   */
+  public boolean passes(String pointLine) {
+    if ((pointLineWhiteList != null && !pointLineWhiteList.matcher(pointLine).matches())
+        || (pointLineBlackList != null && pointLineBlackList.matcher(pointLine).matches())) {
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricWhiteBlackList.java]: The binary operator is usually on the previous line

File: java-lib/src/main/java/com/wavefront/ingester/PickleProtocolDecoder.java [hunk 31]
```diff
@@ -0,0 +1,145 @@
...
+  public void decodeReportPoints(byte[] msg, List<ReportPoint> out, String customerId) {
+    InputStream is = new ByteArrayInputStream(msg);
+    Unpickler unpickler = new Unpickler();
+    Object dataRaw = null;
+    try {
+      dataRaw = unpickler.load(is);
+      if (!(dataRaw instanceof List)) {
+        throw new IllegalArgumentException("unable to unpickle data");
+      }      
+    } catch (final java.io.IOException ioe) {
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/ingester/PickleProtocolDecoder.java]: why the fully qualified class?

File: java-lib/src/main/java/com/wavefront/ingester/PickleProtocolDecoder.java [hunk 32]
```diff
@@ -0,0 +1,145 @@
...
+    this.defaultHostName = hostName;
+    Preconditions.checkNotNull(customSourceTags);
+    this.customSourceTags = customSourceTags;
+    this.metricMangler = mangler;
+  }
+
+  @Override
+  public void decodeReportPoints(byte[] msg, List<ReportPoint> out, String customerId) {
+    InputStream is = new ByteArrayInputStream(msg);
+    Unpickler unpickler = new Unpickler();
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/ingester/PickleProtocolDecoder.java]: I am not sure how heavy this class is but usually you can use a ThreadLocal

File: java-lib/src/main/java/com/wavefront/ingester/PickleProtocolDecoder.java [hunk 33]
```diff
@@ -0,0 +1,145 @@
...
+      dataRaw = unpickler.load(is);
+      if (!(dataRaw instanceof List)) {
+        throw new IllegalArgumentException("unable to unpickle data");
+      }      
+    } catch (final java.io.IOException ioe) {
+      throw new IllegalArgumentException("unable to unpickle data", ioe);
+    }
+
+    // [(path, (timestamp, value)), ...]
+    List<Object[]> data = (List<Object[]>) dataRaw;
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/ingester/PickleProtocolDecoder.java]: if dataRaw nullable?

File: java-lib/src/main/java/com/wavefront/ingester/PickleProtocolDecoder.java [hunk 34]
```diff
@@ -0,0 +1,145 @@
...
+    for (Object[] o : data) {
+      Object[] details = (Object[])o[1];
+      if (details == null || details.length != 2) {
+        logger.warning("Unexpected pickle protocol input");
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/ingester/PickleProtocolDecoder.java]: could this happen a lot (and spam logs?) please include some information perhaps at least the port that received the invalid data perhaps

File: java-lib/src/main/java/com/wavefront/ingester/PickleProtocolDecoder.java [hunk 35]
```diff
@@ -0,0 +1,145 @@
...
+        logger.warning("Unexpected pickle protocol input (timestamp is null)");
+        continue;
+      } else if (details[0] instanceof Double) {
+        ts = ((Double)details[0]).longValue() * 1000;
+      } else if (details[0] instanceof Long) {
+        ts = ((Long)details[0]).longValue() * 1000;
+      } else if (details[0] instanceof Integer) {
+        ts = ((Integer)details[0]).longValue() * 1000;
+      } else {
+        logger.warning("Unexpected pickle protocol input (details[0]: "
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/ingester/PickleProtocolDecoder.java]: method invocation line breaks does not align at the parenthesis

File: java-lib/src/main/java/com/wavefront/ingester/PickleProtocolDecoder.java [hunk 36]
```diff
@@ -0,0 +1,145 @@
...
+        }
+        if (host == null) {
+          host = this.defaultHostName;
+        }
+      }
+      point.setHost(host);
+      point.setTable(customerId);
+      point.setTimestamp(ts);
+      point.setValue(value);
+      point.setAnnotations(new HashMap<String, String>());
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/ingester/PickleProtocolDecoder.java]: Collections.emptyMap()

File: java-lib/src/main/java/com/wavefront/ingester/StreamIngester.java [hunk 37]
```diff
@@ -0,0 +1,69 @@
...
+      b.group(new NioEventLoopGroup(1), new NioEventLoopGroup())
+          .channel(NioServerSocketChannel.class)
+          .option(ChannelOption.SO_BACKLOG, 1024)
+          .localAddress(listeningPort)
+          .childHandler(new ChannelInitializer<SocketChannel>() {
+            @Override
+            public void initChannel(SocketChannel ch) throws Exception {
+              ChannelPipeline pipeline = ch.pipeline();
+              pipeline.addLast("frame decoder", frameDecoderFactory.getDecoder());
+              pipeline.addLast("To Byte Array", new ByteArrayDecoder());
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/ingester/StreamIngester.java]: names are a bit inconsistent

File: java-lib/src/test/java/com/wavefront/common/MetricManglerTest.java [hunk 38]
```diff
@@ -0,0 +1,67 @@
...
+
+  @Test
+  public void testSourceMetricParsing() {
+    String testInput = "hosts.sjc123.cpu.loadavg.1m";
+    String expectedOutput = "cpu.loadavg.1m";
+
+    {
+      MetricMangler mangler = new MetricMangler("2", "", "1");
+      MetricMangler.MetricComponents c = mangler.extractComponents(testInput);
+      Assert.assertEquals(expectedOutput, c.metric);
```
Comment (panghy) [inline java-lib/src/test/java/com/wavefront/common/MetricManglerTest.java]: our code typically use static imports for junit methods

File: proxy/src/main/java/com/wavefront/agent/ChannelByteArrayHandler.java [hunk 39]
```diff
@@ -0,0 +1,108 @@
...
+import java.util.Arrays;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.annotation.Nullable;
+
+
+/**
+ * Channel handler for byte array data.
```
Comment (panghy) [inline proxy/src/main/java/com/wavefront/agent/ChannelByteArrayHandler.java]: author

File: proxy/src/main/java/com/wavefront/agent/PushAgent.java [hunk 40]
```diff
@@ -72,6 +77,14 @@ protected void startListeners() {
         }
       }
     }
+    if (picklePorts != null) {
+      for (String strPort : picklePorts.split(",")) {
```
Comment (panghy) [inline proxy/src/main/java/com/wavefront/agent/PushAgent.java]: Splitter.on(...)....

File: proxy/src/main/java/com/wavefront/agent/PushAgent.java [hunk 41]
```diff
@@ -121,6 +134,32 @@ protected void startOpenTsdbListener(String strPort) {
     new Thread(new Ingester(graphiteHandler, port)).start();
   }
 
+  protected void startPickleListener(String strPort, GraphiteFormatter formatter) {
+    int port = Integer.parseInt(strPort);
+    
+    // Set up a custom handler
+    ChannelHandler handler = new ChannelByteArrayHandler(new PickleProtocolDecoder("unknown", customSourceTags, formatter.getMetricMangler()), port, prefix, pushLogLevel, pushValidationLevel, pushFlushInterval, pushBlockedSamples, getFlushTasks(port), whitelistRegex, blacklistRegex);
```
Comment (panghy) [inline proxy/src/main/java/com/wavefront/agent/PushAgent.java]: line too long

File: proxy/src/test/java/com/wavefront/agent/formatter/GraphiteFormatterTest.java [hunk 42]
```diff
@@ -70,4 +70,54 @@ public void testCollectdGraphiteParsing() {
...
+    String expected1 = "hosts.collectd.cpu.loadavg.1m 40 source=host1";
+    String expected2 = "hosts.collectd.cpu.loadavg.1m 7 1459527231 source=host1";
+
+    Assert.assertEquals(expected1, formatter.apply(testString1));
+    Assert.assertEquals(expected2, formatter.apply(testString2));
+
+    // <= 0 number
+    threwException = false;
+    try {
+      formatter = new GraphiteFormatter(format, delimiter, "0");
```
Comment (panghy) [inline proxy/src/test/java/com/wavefront/agent/formatter/GraphiteFormatterTest.java]: typical paradigm is to add a fail() after line 116

File: java-lib/src/main/java/com/wavefront/common/MetricMangler.java [hunk 43]
```diff
@@ -1,57 +1,58 @@
...
- *     - extracted from one or more components of the metric name
- *       (where each component is separated by a '.')
- *     - allow characters to be optionally replaced in the components extracted
- *       as source name with '.'
- *   - metric name:
- *     - remove components (in addition to the source name)
- *       (this allows things like 'hosts.sjc1234.cpu.loadavg.1m' to get
- *        changed to cpu.loadavg.1m after extracting sjc1234)
- *
+ * Handles updating the metric and source names by extracting components from the metric name. There
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/common/MetricMangler.java]: The original formatting is better but you just need to use javadoc (html) syntax

Comment (wavefront-mike) [inline java-lib/src/main/java/com/wavefront/common/MetricMangler.java]: Ah yes.  I reformatted using intellij.  Looks like it didn't like my non-html formatting.  I'll fix it so it's using html.

**Pull request merged**